### PR TITLE
Add favourite feature

### LINKED ISSUE #22: Favorite feature
As a user, I want to favorite a Client so that I can separate Clients based on whose information I frequent the most (favorited) and those that are not.

### PR DESCRIPTION
**Pull request by hsiaojietng**
Allows the user to `favourite` or `unfavourite `a `Person`. 

Person who is a 'favourite' will appear in the Favourite Window which can be displayed by clicking on 'Favourited' in the menu of 'File'

Includes unit tests that checks if a `Person `is favourited or unfavourited successfully. Also tests for error when an invalid index of Person is supplied when favouriting a `Person`.

Closes #22

### CODE REVIEW

File: src/main/java/seedu/address/ui/MainWindow.java [hunk 1]
```diff
@@ -163,6 +165,14 @@ private void handleExit() {
         primaryStage.hide();
     }
 
+    /**
+     * Opens up FavouriteWindow by setting up the fxml scene and opening it
+     */
+    @FXML
+    private void handleFavourite(ActionEvent event) {
+        favouriteWindow.show();
+    }
```
Comment (howardwhw2) [inline src/main/java/seedu/address/ui/MainWindow.java:174]: Perhaps you might want to add checking for if the favorite window is open already? Maybe something similar to what is being done for the help window
```
   @FXML
    private void handleFavourite(ActionEvent event) {
        if (!favoriteWindow.isShowing()) {
            favouriteWindow.show();
        } else {
            favouriteWindow.focus();
        }
    }
```

**Review by howardwhw2 — APPROVED** [review]
No issues found with the UI apart from a QOL change for consideration.

**Comment by waffledood** [comment]
Hi Jiet, seems like your commit to correct the failing unit test is still unsuccessful. The CI script runs the same gradle script (which runs the unit tests amongst other things) available locally. Perhaps you can run the tests locally before pushing to remote? 

Picture below shows how you can run the unit tests 

image

File: src/main/java/seedu/address/ui/MainWindow.java (hunk 1 above)
Comment (hsiaojietng) [reply src/main/java/seedu/address/ui/MainWindow.java:174]: Thanks @howardwhw2, I have made the necessary changes. By the way, I am still trying to solve the CI workflow because all my tests passed locally, but it does not on Github and I am not sure why.

Comment (waffledood) [reply src/main/java/seedu/address/ui/MainWindow.java]: Hmm that's strange, did you remove/edit any of the unit tests locally?

If not, could you try running that 1 specific unit test your feature is failing? `execute_allFieldsSpecifiedUnfilteredList_success()` Test of `EditCommandTest`

Comment (hsiaojietng) [reply src/main/java/seedu/address/ui/MainWindow.java]: I tested that specific unit test and it passed
image

File: src/main/java/seedu/address/logic/commands/FavouriteCommand.java [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+ * Favourites an existing client in the address book by their index number
+ */
+public class FavouriteCommand extends Command {
+    public static final String COMMAND_WORD = "favourite";
+    public static final String MESSAGE_FAVOURITE_PERSON_SUCCESS =
+            "Favourited Client %1$s! Check if he/she exists in the Favourite Window!";
+    public static final String MESSAGE_UNFAVOURITE_PERSON_SUCCESS =
+            "Unfavourited Client %1$s! Check that he/she is removed from the Favourite Window!";
+    public static final String MESSAGE_USAGE = COMMAND_WORD
+            + ": Favourites a client selected\n"
```
Comment (waffledood) [inline src/main/java/seedu/address/logic/commands/FavouriteCommand.java:24]: Minor cosmetic detail but I noticed that if the String `COMMAND_WORD:  Favourites a client selected` is very long, then appending a `\n` to break up sentences would be necessary. In this case, a newline break mid sentence `"Favorites a client selected by the index number used in the last client listing."` seems odd. 

For example, `MESSAGE_USAGE` in `AddCommand` is quite short `add: Adds a person to the address book. Parameters: n/NAME p/PHONE e/EMAIL a/ADDRESS t/USERTYPE...` hence the inclusion of `\n` only after `Parameters: …`

File: src/main/java/seedu/address/model/person/UniquePersonList.java [hunk 3]
```diff
@@ -97,13 +97,24 @@ public void setPersons(List<Person> persons) {
...
+    }
+
     /**
      * Returns the backing list as an unmodifiable {@code ObservableList}.
      */
     public ObservableList<Person> asUnmodifiableObservableList() {
         return internalUnmodifiableList;
     }
 
+
```
Comment (waffledood) [inline src/main/java/seedu/address/model/person/UniquePersonList.java:117]: additional line break?

File: src/main/java/seedu/address/storage/JsonAdaptedPerson.java [hunk 4]
```diff
@@ -103,6 +107,17 @@ public Person toModelType() throws IllegalValueException {
...
+        if (favourite == null) {
+            throw new IllegalValueException(String.format(MISSING_FIELD_MESSAGE_FORMAT,
+                    Favourite.class.getSimpleName()));
+        }
+        final Favourite modelFavourite;
+        if (favourite.equals("Favourited")) {
+            modelFavourite = new Favourite(true);
+        } else {
+            modelFavourite = new Favourite(false);
+        }
```
Comment (waffledood) [inline src/main/java/seedu/address/storage/JsonAdaptedPerson.java:119]: Would it be possible to have something like `Favourite.isFavourited(favourite)` similar to how the other classes `Email`, `Address` in the same `toModelType()` method of `Person` class did it? 

That way, all functionalities of `Favourite` are contained within the class itself

File: src/main/java/seedu/address/ui/MainWindow.java [hunk 5]
```diff
@@ -163,6 +165,18 @@ private void handleExit() {
...
+     */
+    @FXML
+    private void handleFavourite(ActionEvent event) {
+        if (!favouriteWindow.getRoot().isShowing()) {
+            favouriteWindow.show();
+        } else {
+            favouriteWindow.getRoot().requestFocus();
+        }
+    }
+
```
Comment (waffledood) [inline src/main/java/seedu/address/ui/MainWindow.java:179]: good implementation of Howard's recommendation! 👍

File: src/test/java/seedu/address/testutil/PersonBuilder.java [hunk 6]
```diff
@@ -94,8 +98,16 @@ public PersonBuilder withEmail(String email) {
         return this;
     }
 
+    /**
+     * Sets the {@code Email} of the {@code Person} that we are building.
+     */
```
Comment (waffledood) [inline src/test/java/seedu/address/testutil/PersonBuilder.java:103]: did you mean `Sets the {@code Favourite}`?

File: build.gradle [hunk 7]
```diff
@@ -34,6 +34,10 @@ task coverage(type: JacocoReport) {
             fileTree(dir: it, exclude: ['**/*.jar'])
         })
     }
+    tasks.withType(Test) {
+        scanForTestClasses = false
+        include "**/*Test*.class"
+    }
```
Comment (waffledood) [inline build.gradle:40]: why was this added?

**Review by waffledood — APPROVED** [review]
LGTM, well-written clean code! 

Just some queries & feedback

File: src/main/java/seedu/address/ui/MainWindow.java (hunk 1 above)
Comment (waffledood) [reply src/main/java/seedu/address/ui/MainWindow.java]: any chance your tests were not set up properly? you could perhaps run through the setup again on the cs2103 website

File: src/main/java/seedu/address/logic/commands/FavouriteCommand.java (hunk 2 above)
Comment (hsiaojietng) [reply src/main/java/seedu/address/logic/commands/FavouriteCommand.java:24]: Thanks for the suggestion. I have made necessary changes.

File: src/main/java/seedu/address/model/person/Person.java [hunk 8]
```diff
@@ -21,20 +21,39 @@
...
+        this.tags.addAll(tags);
+    }
+
     /**
      * Every field must be present and not null.
+     * This constructor is used for adding a new Client, thus default status is unfavourited(false)
      */
     public Person(Name name, Phone phone, Email email, Address address, Optional<Property> property, Set<Tag> tags) {
-        requireAllNonNull(name, phone, email, address, property, tags);
+        requireAllNonNull(name, phone, email, address, tags);
```
Comment (Ashuh) [inline src/main/java/seedu/address/model/person/Person.java:52]: Is there any reason for removing `property` from `requireAllNonNull()`?

File: src/main/resources/view/FavouriteWindow.fxml [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+<?import javafx.scene.Scene?>
+<?import javafx.scene.control.Menu?>
+<?import javafx.scene.control.MenuBar?>
+<?import javafx.scene.control.MenuItem?>
+<?import javafx.scene.image.Image?>
+<?import javafx.scene.layout.StackPane?>
+<?import javafx.scene.layout.VBox?>
+<?import javafx.stage.Stage?>
+
+<Stage minHeight="600" minWidth="450" title="Address App" xmlns="http://javafx.com/javafx/17" xmlns:fx="http://javafx.com/fxml/1">
```
Comment (cindyangXQ) [inline src/main/resources/view/FavouriteWindow.fxml:14]: Could change the title to "Favorite Window"

File: src/main/java/seedu/address/storage/JsonAdaptedPerson.java (hunk 4 above)
Comment (hsiaojietng) [reply src/main/java/seedu/address/storage/JsonAdaptedPerson.java:119]: Resolved!

File: build.gradle (hunk 7 above)
Comment (hsiaojietng) [reply build.gradle:40]: Tried this to resolve the CI workflow error but it didnt work, so I will remove it.

File: src/main/resources/view/FavouriteWindow.fxml (hunk 9 above)
Comment (hsiaojietng) [reply src/main/resources/view/FavouriteWindow.fxml:14]: Thanks for the catch!

File: src/main/java/seedu/address/model/person/Person.java [hunk 10]
```diff
@@ -21,20 +21,39 @@
     private final Name name;
     private final Phone phone;
     private final Email email;
+    private Favourite favourite;
```
Comment (Ashuh) [inline src/main/java/seedu/address/model/person/Person.java:24]: Possible to set this as `final`?

File: src/main/java/seedu/address/model/person/Person.java [hunk 11]
```diff
@@ -52,6 +71,18 @@ public Email getEmail() {
...
+        return favourite;
+    }
+
+    /**
+     * Toggles the favourite status of Person
+     */
+    public void toggleFavourite() {
+        boolean toggledStatus = !favourite.getStatus();
+        favourite.setStatus(toggledStatus);
+    }
```
Comment (Ashuh) [inline src/main/java/seedu/address/model/person/Person.java:84]: Will this violate the guarantee that field values are immutable?

**Review by Ashuh — COMMENTED** [review]
Code looks good except for a few issues

File: src/main/java/seedu/address/model/person/Person.java (hunk 8 above)
Comment (hsiaojietng) [reply src/main/java/seedu/address/model/person/Person.java:52]: Thanks for the confirmation, cause this was a newly created constructor that was written before your PR. Will include the `property` field in.

File: src/main/java/seedu/address/model/person/Person.java (hunk 10 above)
Comment (hsiaojietng) [reply src/main/java/seedu/address/model/person/Person.java:24]: Thanks for the suggestion, made the necessary changes

File: src/main/java/seedu/address/model/person/Person.java (hunk 11 above)
Comment (hsiaojietng) [reply src/main/java/seedu/address/model/person/Person.java:84]: I'm not sure if I understand this correctly, but if you mean does it change the value of `Favourite` field of a Person, then yes this field should be mutable since the user is able to 'Favourite' or 'Unfavourite' a Person based on his/her `Favourite `status currently.

Comment (Ashuh) [reply src/main/java/seedu/address/model/person/Person.java:84]: The reason im asking is because the javadoc for the class currently states that all fields are immutable.

So we need to either remove this guarantee or make this method return a new instance of `Person`

Comment (hsiaojietng) [reply src/main/java/seedu/address/model/person/Person.java:84]: I think it would be more logical to change the field of the `Person` rather than return a new instance of `Person`, what do you think?

Comment (Ashuh) [reply src/main/java/seedu/address/model/person/Person.java:84]: Im fine with that but make sure to update the javadoc
